### PR TITLE
Support shiny apps

### PR DESCRIPTION
**Pull request by ryanlovett**
nbrsessionproxy supports the proxying of shiny-server when it is installed. This PR installs shiny-server and the shiny R library. It also makes some unconfigurable shiny paths writable by the notebook user.

When someone visits /shiny/, the user server will proxy to shiny-server. If there is a shiny app in the home directory, shiny will run that app, otherwise it will display shiny's tree listing. Visiting directories containing shiny apps will run them.

I tested this by generating a Dockerfile with --debug --no-build, then manually seeding the R buildpack bits. If there's a better way of doing this please let me know!

### CODE REVIEW

File: repo2docker/buildpacks/r.py [hunk 1]
```diff
@@ -186,6 +204,13 @@ def get_build_scripts(self):
                 """.format(
                     devtools_version=devtools_version,
                     irkernel_version=irkernel_version
+            ),
+            (
+                "${NB_USER}",
+                # Install shiny library
+                r"""
+                R --quiet -e "install.packages('shiny', repos='https://mran.microsoft.com/snapshot/{}', method='libcurl')"
+                """.format(self.checkpoint_date.isoformat())
```
Comment (betatim) [inline repo2docker/buildpacks/r.py]: I think this is the offending line -> delete the last closing `)`

**Comment by betatim** [comment]
This looks great!

Trying to debug where there is a missing `,` or some such that is making travis unhappy.

File: repo2docker/buildpacks/r.py (hunk 1 above)
Comment (betatim) [reply repo2docker/buildpacks/r.py]: Made a PR to your repo https://github.com/ryanlovett/repo2docker/pull/1 (I thought these days repo maintainers can push to the PR branches of others?)

**Comment by betatim** [comment]
This is gonna be awesome.

Is there a example shiny app we can add to https://github.com/binder-examples/r or a new example repo we could add to https://github.com/binder-examples/?

**Comment by ryanlovett** [comment]
Thanks for the merge! I tend to test with this [single file app](https://shiny.rstudio.com/gallery/single-file-shiny-app.html) and the [gallery](https://shiny.rstudio.com/gallery/) is full of nice ones.

RStudio's [shiny-examples](https://github.com/rstudio/shiny-examples/) has them all lined up in a row.

**Comment by betatim** [comment]
The `hello` app works, but https://github.com/rstudio/shiny-examples/tree/master/086-bus-dashboard just gives me "ERROR: An error has occurred. Check your logs or contact the app author for clarification." Is there a way to see the error log somewhere? `/home/jovyan/logs/` is empty :(. I downloaded the whole `086-bus-dashboard` directory and put it in a sub directory of [binder-examples/r](https://github.com/binder-examples/r)

n.b. I had to delete the `index.Rmd` in the root directory to get shiny server to show me the directory listing. I think that is everything working as designed but took me a moment to figure out.

**Comment by ryanlovett** [comment]
Thanks for looking into this. The shiny logs should be in /var/log/shiny-server.log and /var/log/shiny-server/. The first is the shiny-server process and the second contains log files for each app.

The closest thing to jupyter server logs I've found on binder is `cat /proc/1/fd/2` in a jupyter terminal, if I recall correctly.

I'm traveling atm, but I'll try to replicate the 086-bus-dashboard issue.

**Comment by betatim** [comment]
The log at `/var/log/shiny-server.log` is empty and looking at the output you get from running `repo2docker .` locally is the following:
```
[E 05:21:09.481 NotebookApp] {
      "Host": "0.0.0.0:57883",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      "Accept-Language": "en-US,en;q=0.5",
      "Accept-Encoding": "gzip, deflate",
      "Referer": "http://0.0.0.0:57883/shiny/",
      "Cookie": "username-0-0-0-0-50436=\"2|1:0|10:1527340252|22:username-0-0-0-0-50436|44:NDdhYzY1Yzg0ZmM4NDBhN2E5NDFkNmQwMjFiNjY4MDg=|db6f8c9eeb3c844611d7948e37777e706ac8cf3ef263887e3eb9161af467f594\"; _xsrf=2|da088ef2|49aa05d4ce69585f7b8151a9e195e1ab|1527340252; username-0-0-0-0-51341=\"2|1:0|10:1527345393|22:username-0-0-0-0-51341|44:NmU4YzBkYWY2MDY4NGQ1Zjg0Y2NlNGE0NzA1M2Y3NWI=|83fd82c9b5488d1186fb85f59a9d000556b41e28b8bcda0505ec01d6478320d2\"; username-0-0-0-0-57883=\"2|1:0|10:1527743676|22:username-0-0-0-0-57883|44:NmFhOTJjMTRmNjI5NGMyZGFjZjZlODEwN2IzYzUyZGE=|c9ce84c4dfd71ba4dead686eecf20b723190c5014cc8e1a07a41665e8cb8ef42\"",
      "Dnt": "1",
      "Connection": "keep-alive",
      "Upgrade-Insecure-Requests": "1",
      "Cache-Control": "max-age=0"
    }
[E 05:21:09.488 NotebookApp] 500 GET /shiny/bus-dashboard/ (172.17.0.1) 1180.52ms referer=http://0.0.0.0:57883/shiny/
```

Where else could I look for why this fails?

**Comment by betatim** [comment]
The problem was missing dependencies for the shiny app. After finding and adding them to `install.R` you can visit `/shiny/bus-dashboard` and get the bus dashboard.

Finding out that this was the problem was a bit tricky. In the end I opened the `server.R` in RStudio and clicked "launch App" there. This gave me a window with an error message that said "missing dependency XXX". Didn't manage to find that in the logs unfortunately.

**Pull request merged**